### PR TITLE
Add a timeout to the `@_loop` recipe

### LINKED ISSUE #562: `just init` sometimes gets stuck when waiting for indices
## Description

Sometimes, quite randomly, `just init` gets stuck waiting for indices to be ready.

```text
curl -X POST -H 'Content-Type: application/json' -d '{"model": "image", "action": "LOAD_TEST_DATA"}' 'http://localhost:8001/task'
{"message": "Successfully scheduled task", "task_id": "ee930da1-5013-4133-b9e7-39c02a669ef3", "status_check": "http://localhost:8001/task/ee930da1-5013-4133-b9e7-39c02a669ef3"}
curl -X POST -H 'Content-Type: application/json' -d '{"model": "audio", "action": "LOAD_TEST_DATA"}' 'http://localhost:8001/task'
{"message": "Successfully scheduled task", "task_id": "e1fa71af-9503-4949-8ce3-039302e83188", "status_check": "http://localhost:8001/task/e1fa71af-9503-4949-8ce3-039302e83188"}
curl -X POST -H 'Content-Type: application/json' -d '{"model": "image", "action": "INGEST_UPSTREAM"}' 'http://localhost:8001/task'
{"message": "Successfully scheduled task", "task_id": "4f1b0c36-603e-4e21-93e6-adb1173e576c", "status_check": "http://localhost:8001/task/4f1b0c36-603e-4e21-93e6-adb1173e576c"}
Waiting for index 'image' to be ready...
Waiting for index 'image' to be ready...
Waiting for index 'image' to be ready...
Waiting for index 'image' to be ready...
Waiting for index 'image' to be ready...
Waiting for index 'image' to be ready...
Waiting for index 'image' to be ready...
Waiting for index 'image' to be ready...
````

This can also break CI sometimes.

## Reproduction

1. Clean up the Docker space.
    ```shell-session
    $ docker compose down -v
    ```
1. Bring up the services and init the database.
    ```shell-session
    $ just init
    ```
1. Repeat till you see it get stuck.

## Expectation
This should not get stuck, or should have timeout and retry functionality.

## Resolution

- [ ] 🙋 I would be interested in resolving this bug.

### PR DESCRIPTION
**Pull request by dhruvkb**
## Description

Prevents Just recipes for waiting to enter into endless loops. This is part of a larger effort to fix #562.

## Testing Instructions

0. Change the condition in any Just recipe that uses `@_loop` to fail.
1. Run the Just recipe.
2. It will timeout after 5m (or the set timeout duration).

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: justfile
Comment (sarayourfriend) [inline justfile:13]: 5 minutes is still a long time! Is that a reasonable amount of time to expect ES to take to boot up? I would have thought like 2 minutes max but maybe it could take a lot longer on smaller CPUs (or even in the action runners).

**Review by sarayourfriend — COMMENTED** [review]
TIL about the `timeout` utility!

Will note that this is _not_ cross-platform.

We could write the script as a Python script instead if we wanted to ensure that: https://github.com/casey/just/#writing-recipes-in-other-languages

**Comment by dhruvkb** [comment]
The `timeout` utility works on Linux and macOS. Considering that the `load_sample_data.sh` is a Bash script, I'm pretty sure we're not supporting Windows as a development environment right now.

**Comment by sarayourfriend** [comment]
That's true, I wonder if we should add further things on the list of stuff to convert to support Windows developers?

Or else we should decide that we don't care about Windows devs (which I don't think is the right answer considering Window's marketshare) and make that explicit in the README and contributor docs.

**Comment by zackkrida** [comment]
I would imagine many devs are using or have access to [WSL](https://docs.microsoft.com/en-us/windows/wsl/about) these days, perhaps we could add some simple docs to suggest it and consider that basic Windows support.

**Comment by dhruvkb** [comment]
Supporting Windows is important. We could definitely put some WSL instructions in the docs but ensuring everything works.

**Comment by sarayourfriend** [comment]
Would it be possible to print the last few lines of docker logs for the elastic container when the operation times out? Might help in debugging failures in CI without having to try reproducing the flakiness locally.

**Comment by AetherUnbound** [comment]
It seems like `timeout` will exit with code 124 if the timeout is hit. Do you think we could use that to echo a message? For instance, if the command's error code is 124, `echo "Command failed after {{ timeout }}"` or something similar. Currently the output isn't super obvious:

```
$ just wait-for-ing
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
error: Recipe `_loop` failed on line 16 with exit code 124
error: Recipe `wait-for-ing` failed on line 134 with exit code 124
```

**Comment by zackkrida** [comment]
@dhruvkb do you need reviews here, or to implement changes? Please feel free to ping the assigned reviewers if the former. Thanks

**Comment by dhruvkb** [comment]
Needs some more work. Will ping reviewers as soon as ready.

**Comment by sarayourfriend** [comment]
@dhruvkb just wanted to check if you needed any support on this PR?

**Comment by dhruvkb** [comment]
@sarayourfriend just need to implement what @AetherUnbound recommended here: https://github.com/WordPress/openverse-api/pull/564#issuecomment-1067387898

**Review by AetherUnbound — CHANGES_REQUESTED** [review]
Hmm, doesn't seem like the current implementation works on linux :( 
```
$ just wait-for-ing
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
Waiting for the ingestion-server to be healthy...
sh: 1: [[: not found
```

**Comment by AetherUnbound** [comment]
Setting the following at the top of the `justfile` worked for me though! `set shell := ["bash", "-uc"]`

```
Waiting for the ingestion-server to be healthy...
Timed out
error: Recipe `_loop` failed on line 22 with exit code 124
error: Recipe `wait-for-ing` failed on line 145 with exit code 124
```

**Comment by dhruvkb** [comment]
@AetherUnbound are you using a shell other than `bash` as the default? I'm guessing based on our error message that says:
```
sh: 1: [[: not found
```

**Comment by AetherUnbound** [comment]
I'm not, `bash` is what I'm using now! :woman_shrugging: I take it this doesn't error out for you when it's run? If you add `echo $0` at the beginning of the loop recipe, what do you get? :open_mouth:

**Comment by dhruvkb** [comment]
I'm getting `sh` too when I add the `echo $0`. But it turns out macOS was being sneaky.
Screenshot 2022-04-06 at 6 08 48 AM

I've updated the script to use `[` instead of `[[` in ed96453 so it should work consistently across shells now without needed to specify one at the top of the `justfile`.

**Comment by AetherUnbound** [comment]
Whaaaaaa :open_mouth: how wild! Well this works for me now, thanks Dhruv!

Fun side note: I found this line while reading the man page for `sh` (apparently `sh` doesn't have a `-version` flag):
> ### Description
> Sh is the standard command interpreter for the system. The current version of sh is **in the process of being changed to conform with the POSIX 1003.2 and 1003.2a specifications for the shell**.  _(emphasis mine)_

I wanted to know how long `sh` has been "in the process of changing its specification", and it turns out that the [oldest man page I can find for `sh` from 1999](https://linux.die.net/man/1/ash) _also has this line_ :joy: ahhh, software.

**Comment by dhruvkb** [comment]
Thanks, just need one more approval. @sarayourfriend or @krysal if you have a few moments, please give it a look.

**Review by zackkrida — APPROVED** [review]
lgtm!

**Pull request merged**
